### ISSUE TITLE
[Question]: How could we close/update previous PR comments ?

### ISSUE THREAD

**Issue by jogelin** [joachimdalen/AzureDevOpsExtensions#2]
Hello,

I am using your `PullRequestUtils` to publish a comment on each build. I am using it to publish the URL of the deployed build artifact. So each time a build is done, a new unique URL is generated. 

But I don't know how can I remove/update the previous comment/url of a previous build ? 
Would it be possible to have an id of the comment to be able to edit/delete it on a new build ?

I also tried the `skipIfCommentExists` options but it doesn't work because I believe you check the equality right ?

**Comment by joachimdalen** [comment]
Hi there :wave: 

Unfortunately the task only supports creating comments, not updating or deleting them. Adding these features would extend the use cases for the task, so I think it is a good idea to support it :tada: 

Your assumptions about the `skipIfCommentExists` option is correct. It only checks the content of the comment, so with the content even being one character off it would still post the comment. Appending some sort of identifier/name to the comment that identifies the step and iteration in the pipeline (and ignoring the actual content of the comment) does not sound like a bad idea. Then use cases like the one you describe would be better supported.

I have some time today, so I will look into adding/improving this.

**Comment by jogelin** [comment]
@joachimdalen thanks for your quick answer !!

How could I support you with new features? 
I don't know how azure devops extension is written, I'll have a look.

I am asking this question because I think, for my use case, the best approach isn't the publication of comments. The update of the PR description would be better. 

I saw there is an API to that I could use to do that so I was ready to write an external script for that but a yaml task would be better.

**Comment by joachimdalen** [comment]
> How could I support you with new features? I don't know how azure devops extension is written, I'll have a look.

Currently there isn't a good way. I'm working on some cleanup to the code and planning on open sourcing them when complete. The best way to help so far is to just give feedback/questions like you have. It allows me to see users perspective and improve it for everyone, so thanks :smile: 

> I am asking this question because I think, for my use case, the best approach isn't the publication of comments. The update of the PR description would be better.

Yeah, I see that in some cases actually updating the description would be better than posting/updating comments. Actually creating a task to do that wouldn't be to hard by just using the API endpoint: [Pull Requests - Update](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/pull%20requests/update?view=azure-devops-rest-5.1).  **I need to double check permissions in Azure DevOps before saying for certain that the pipeline can update a description created by someone else**

The issue here is how to handle the different states and how to prevent the pipeline task from actually overwriting the creators description. When you create a script for a single pipeline it is easy enough because you know the format of the text you want to update, but to generalize this idea is a bit harder.

Looking at both the API and the PR in Azure DevOps I can't find any metadata that would allow us know when the description was last updated either. This means that if we somehow store the original description (e.g in a pipeline artifact) we wouldn't know when to re-fetch it from the API. 

One possibility would be to do something like appending a markdown comment to the end of the pull request on the first run, and on sequential runs it would replace anything below that. 

Let me give an example:
```
## First run 
This would be the description for the first run
```
When the pipeline runs for the first iteration on the pull request, it will be updated to:

```
## First run 
This would be the description for the first run

[//]: # (pull-request-description-updater - Anything below this line will be deleted on next pipeline run)
Your content from the task goes here
```
`[//]: # (pull-request-description-updater - Anything below this line will be deleted on next pipeline run)` - This line would not be visible to user except for in edit mode (it's a markdown comment)

When the pipeline runs again, it would delete everything below the  `[//]: # (pull-request-description-updater - Anything below this line will be deleted on next pipeline run)` line and replace it with the updated content.

Would something like this help you solve the initial problem?

**Comment by jogelin** [comment]
@joachimdalen 
I think it is a good idea to keep a section hidden (like a comment) in the description with the metadata. It could be used then by the extension. 

I was thinking the same for the comments to identify them even if the content is different. We could provide an identifier and the`skipIfCommentExists` could use it if exists. Or have a way to overwrite an existing comment with a new content

For the existing PR description, I would propose to have a parameter that would specify the type of merge you want to use. 
For example:
```
steps:
  - task: PullRequestDescription@0
    inputs:
      content: 'This is my new content'
      merge: 'append'
```
where:
* `append`: add content to the end
* `prepend`: add content to the begin
* `replace`: replace the content
* `inject`: inject content at a specific place where you find a specific tag in the content (this one feels less easy to put in place)

**Comment by joachimdalen** [comment]
Yes, I think we are onto something. I will work some on it and keep you posted later.

**Comment by joachimdalen** [comment]
Hi again @jogelin 

Just a little update. I said previously that I had to double check permissions in Azure DevOps before saying for certain that the pipeline can update a description created by someone else. This can indeed be done. I've so far created a development version of `PullRequestDescription` that supports `replace` and `append`. Using markdown comments seems to be working good so far. I will also implement `prepend` before publishing it. Also working on improving the `PullRequestComments` task. 

I should have an update of the extension published for you to try in the next few days.

**Comment by joachimdalen** [comment]
Most things are in place now. I just need to do some more testing. Aiming for a release so you can use it sometime tomorrow :rocket:

**Comment by joachimdalen** [comment]
Hi @jogelin 

I just published version 0.4.2 of the extension. This includes functionality for updating comments and their statuses. The `skipIfCommentExists` feature has also been improved.

`PullRequestDescription` is the new task, and allows for updating the description of a pull request. So far it supports `replace` and `append`. I will continue to work on `inject` and `prepend`, these will come in a later release.

Please see the [documentation on the extension page](https://marketplace.visualstudio.com/items?itemName=joachimdalen.pull-request-utils) for usage.

Thank you for your input on this :+1: 

Can you try out the new version and see if it works in a way that solves your initial problem?

**Comment by jogelin** [comment]
@joachimdalen Thx a lot, it is working fine! 

I don't think you should too much time on inject or prepend until nobody has the use case.

I keep you informed if I find any issues/improvements.
